Dockerfile

```diff
@@ -1,98 +1,99 @@
 RUN pip3 install \
   beautifulsoup4==4.7.1 \
   lxml==4.3.1 \
   pycodestyle==2.5.0
 
-# Install ruby deps with bundler to make things more standard for Ruby folks.
+
+FROM base AS ruby_deps
+RUN install_packages \
+  bundler \
+    # Fetches ruby dependencies
+  ruby-dev make cmake gcc libc-dev patch
+    # Required to compile some of the native dependencies
 RUN bundle config --global silence_root_warning 1
 COPY Gemfile* /
+# --frozen forces us to regenerate Gemfile.lock locally before using it in
+# docker which lets us lock the versions in place.
 RUN bundle install --binstubs --system --frozen
 COPY .docker/asciidoctor_2_0_10.patch /
 RUN cd /var/lib/gems/2.5.0/gems/asciidoctor-2.0.10 && patch -p1 < /asciidoctor_2_0_10.patch
-# --frozen forces us to regenerate Gemfile.lock locally before using it in
-# docker which is important because we need Gemfile.lock to lock the gems to a
-# consistent version and we can't rely on running bundler in docker to update
-# it because we can't copy from the image to the host machine while building
-# the image.
 
+
+FROM base AS node_deps
+COPY .docker/apt/keys/yarn.gpg /
+RUN apt-key add /yarn.gpg
+COPY .docker/apt/sources.list.d/yarn.list /etc/apt/sources.list.d/
+RUN install_packages yarn
 COPY package.json /
 COPY yarn.lock /
 ENV YARN_CACHE_FOLDER=/tmp/.yarn-cache
+# --frozen-lockfile forces us to regenerate yarn.lock locally before using it
+# in docker which lets us lock the versions in place.
 RUN yarn install --frozen-lockfile
+
+
+FROM base AS final
+LABEL MAINTAINERS="Nik Everett <nik@elastic.co>"
+RUN install_packages \
+  git \
+    # Clone source repositories and commit to destination repositories
+  libnss-wrapper \
+    # Used to clean up user id differences in the docker image.
+  libxml-libxml-perl \
+    # Parses the sitemap
+  libxml2-utils \
+    # Validates the docsbook xml
+  make \
+    # Used by the tests
+  nginx \
+    # Serves docs during tests and when the container is used for "preview" or
+    # "air gapped" docs
+  openssh-client \
+    # Used by git
+  openssh-server \
+    # Used to forward git authentication to the image on OSX
+  perl-base \
+    # The "glue" of the docs build is written in perl
+  python \
+    # AKA python2. Runs AsciiDoc but *not* Asciidoctor.
+  xsltproc
+    # Converts the docbook xml into html
+COPY --from=node_deps /node_modules /node_modules
+COPY --from=py_deps /usr/local/lib/python3.7/dist-packages /usr/local/lib/python3.7/dist-packages
+COPY --from=py_deps /usr/local/bin/pycodestyle /usr/local/bin/pycodestyle
+COPY --from=ruby_deps /var/lib/gems /var/lib/gems
+COPY --from=ruby_deps /usr/local/bin/asciidoctor /usr/local/bin/asciidoctor
+COPY --from=ruby_deps /usr/local/bin/rspec /usr/local/bin/rspec
+COPY --from=ruby_deps /usr/local/bin/rubocop /usr/local/bin/rubocop
+
+# We mount these directories with tmpfs so we can write to them so they
+# have to be empty. So we delete them.
+RUN rm -rf /var/log/nginx && rm -rf /run/nginx
```

Comment:
you could shave few more MB off by running `apt-get clean` to remove downloaded packages from cache.

Reply:
I believe `install_package` does that. Which is actually a problem because we run it a few times so it runs `apt-get update` and clears that cache over and over again.